### ISSUE TITLE
Clarification - variable length CID encoding MUST be deterministic

### ISSUE THREAD

**Issue by boaks** [tlswg/dtls-conn-id#76]
The discussion in the [IETF TLS mailing list](https://mailarchive.ietf.org/arch/browse/tls/?q=AD%20review%20of%20draft-ietf-tls-dtls-connection-id-07) contains many examples with variable length CIDs, where the encoding is assumed to be ambiguous.

In my opinion, the encoding MUST be deterministic/unique.

Any different opinions?

**Comment by ad-l** [comment]
See section III.D of https://eprint.iacr.org/2020/114.pdf for an illustration on QUIC. This can matter if there is any parsing or partial processing between the network representation and the cryptographic representation (what is hashed, signed, MACed, or used as AAD), such as QUIC header protection or cTLS compressed encoding. If any such re-encoding is done, it must be non-malleable and have strong prefix property (see definitions 1 and 2 in https://antoine.delignat-lavaud.fr/doc/usenix19.pdf).

Note that we prove in the Usenix paper than any syntax written the TLS message presentation syntax has these properties assuming you don't write untagged unions. https://github.com/project-everest/everparse is a tool that can check if the message format grammar has this property automatically. For QUIC this is only the case if the canonical (minimal) representation is used; there is a formal proof of the fact here https://github.com/project-everest/everquic-crypto/blob/master/src/QUIC.Spec.VarInt.fsti

**Comment by boaks** [comment]
> This can matter if there is any parsing or partial processing between the network representation and the cryptographic representation

Isn't the only item, which may be the result of a decoding, that cid-length?

The current idea is to assume generally, that the CID is opaque, in order to have very simple implementation on the other's peer side. That other peer only takes the CID received in the hello_extension and applies that to the dtls-cid-records (and use the adapted MAC).

With that only the emitting peer, knows about the details of encoding the CID. That includes, if, at all, a variable length is chosen. Therefore, only that emitting peer may be faced of such issues. For me, that `strong prefix property` must be a property of the CID encoding, not the MAC.

**Comment by jsalowey** [comment]
I think the basic problem would be resolved by switching the order of the cid_length and the cid in the MAC.  However, I agree it would be better to make the length part of the encoding.  There is a mix between implicit and explicit handling for the CID,  this makes it hard to reason about and there may be other problems lurking about.

**Comment by boaks** [comment]
> I think the basic problem would be resolved by switching the order of the cid_length and the cid in the MAC. 

In my opinion, it doesn't. 
All the sayings so far about the cid-length are more analogies from cases, where that cid-length is on the wire (e.g. QUIC). And then, it's about the position also on the wire, and with that, the position in the MAC.
So far, I haven't read something about the special way of this draft. 
That special way may be broken, but also that isn't something, someone has written so far. 

> However, I agree it would be better to make the length part of the encoding.

Currently it's optional to use a variable length CID. 
And only the emitting peer is affected. The current definition enables a kind of "asymmetric implementations", a simple/slim for constraint peers and a flexible/fat for powerful peers. That reflects in my opinion the most common current requirement:

- small devices with "dynamic/temporary" 5-tuples (maybe caused by NATs)
- server with static/fix 5-tuple

For the constrained peers it's simply copies the received cid and apply the new MAC. No details about the cid nor the optional variable length encoding is required on that peer. 

For the powerful peer, it's a implementation decision (e.g. Californium defines a java interface to handle with custom cid encodings, currently mainly used for the cid-cluster-support. see [ConnectionIdGenerator](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ConnectionIdGenerator.java) ).
If the implementation wants to use a variable length CID, it's the responsibility of that implementation, to encode the CID with variable length in a "injective" way. But to define that encoding within this RFC, may be the wrong approach. It's more to require that it's "injective" (hope, I use the right term), then to specify the details. One may use a preceding byte, other just some bits, because the implementation may be limited to use only up to 16 bytes. 

> There is a mix between implicit and explicit handling for the CID, this makes it hard to reason about
      
In my opinion, It's not the CID itself, the handling of that is defined pretty well.
It may be the handling of the cid-length, assuming "variable length CIDs". And here, my feeling is more and more, that analogies are considered to fast to apply here as well.
  
>  and there may be other problems lurking about.

Yes, that's my biggest fear. 
Therefore I would very welcome a short statement from @ekr about his opinion on that topic.

**Comment by boaks** [comment]
From [my e-mail](https://mailarchive.ietf.org/arch/msg/tls/WV1_qyZ4ZqqmgpZEBp_dmSLAbvM/) with some corrections:

- a peer chose to emit only CIDs with the same static length. In that case, the cid-length will always be the same (cid-length is not on the wire). It can't be injected, except someone injects the executed program code itself. For that case, the cid-length seems to be obsolete in the MAC (but doesn't hurt). 

- a peer chose to emit CIDs with variable length. The emitting peer has to chose a encoding for such CID with variable length, that the peer itself is able to decode it. In this case the cid-length is on the wire. In my opinion it's extremely important, to chose a encoding, which could not be interpreted in different ways. But, if the encoding is "injective", then the "logical duplicated" cid-length is also obsolete in the MAC (but doesn't hurt, because it can't be injected, if the encoding is injective). 

In both cases the other peer is intended just to add that received cid in the hello extension to the dtls-records and apply the MAC.

**Comment by boaks** [comment]
Just to mention:

In my opinion, not recommending "variable length CIDs" would be also a way to go.

**Comment by kaduk** [comment]
> The emitting peer has to chose a encoding for such CID with variable length, that the peer itself is able to decode it. [...]

Yes, exactly (and I agree with you on the "extremely important" part).

Some of the previous discussion here does make me want to think a bit harder, though, about the role of the two peers and where the injectivity of mapping is needed. Right now, the (MAC) verifier is the CID issuer, so it knows the internal structure to any variable-length CID and can trivially invert the mapping from protocol fields to MAC input, so that seems okay with or without an explicit CID length in the MAC input.  However, the party that is generating the MAC is not expected to know any internal structure to the CID, and has to blindly trust the CID value (and length) provided by the peer during the handshake.  The tls12_cid-type DTLSCiphertext is used starting with the client Finished (recalling that this is TLS 1.2, which is ... less nice than TLS 1.3), which is before the server Finished or any other key confirmation from the server.  This seems to imply a potential attack where the ServerHello is tampered with and the client produces a Finished using the "wrong" (tampered) CID.  With the information available to the client it cannot invert the mapping from protocol fields to MAC input and it seems possible that it can be convinced to generate a MAC using a CID that has been tampered to make use of this malleability (and the attacker might then devise a way to use the computed MAC with a different plaintext, without knowing the MAC key.  This makes me uncomfortable with omitting the CID length entirely, even though the specific example of malleability that we know about would involve a Finished body that is not the expected length and thus ought to be detected/prevented by one or both legitimate endpoints.

**Comment by boaks** [comment]
> This seems to imply a potential attack where the ServerHello is tampered with and the client produces a Finished using the "wrong" (tampered) CID.

If the cid in the hello extension is considered to be tampered, I can't see, that including the cid-length in the MAC will be of any help, that cid-length may be tampered as well, or?. In my opinion, [RFC 5246 - FINISH](https://tools.ietf.org/html/rfc5246#section-7.4.9) includes a hash over all messages, intended to protect from such modifications. If that client FINISH fails, the generated keys used for this handshake are also devaluated. The next handshake will result in new keys.

In my opinion, assuming that handshake messages are tampered, seems to be a general topic, not really related to include the cid-length into the MAC.

**Comment by kaduk** [comment]
Sorry in advance for being pithy -- I should really be asleep already.
Anyway, if a Dolev-Yao attacker can induce the DTLS client to produce a MAC value that is seen by the server as valid for a message different than the message the client intended to produce a MAC for, this MAC construction is flawed and needs to be replaced.  Whether or not we think that other parts of the handshake would detect the usage of such a modified message is not relevant.

**Comment by boaks** [comment]
In my opinion, you relate an idea of an attack with the assumption, using the cid-length in the right position in the MAC will solve it.  So far, you didn't show, how such a cid-length would help to overcome that attack. 
I'm not sure, if constructing cases, and applying not related partial solutions will help.

**Comment by jsalowey** [comment]
It would be OK to fix both the MAC and the message/CID encoding.   I think that could be the path forward to resolving this issue.

**Comment by boaks** [comment]
Agreed.
(I would prefer, to remove the `cid-length` from the MAC to fix it.)

**Comment by jsalowey** [comment]
Would you also change the CID format to include length of the CID?

**Comment by boaks** [comment]
Currently the draft has no CID format, its opaque.
I would not add a format, which makes an encoded length mandatory. But I would add the constraint, that, if such a "variable length CID" is to be used, it MUST use a injective encoding and add an example, for such a encoding, with the first byte a length.

**Comment by kaduk** [comment]
I am worried that "just" requiring the CID format to be self-framing (i.e., have an internal length) does not provide all the properties that we need.  It ensures that the MAC verifier knows unambiguously what it is being requested to verify, yes, but the party generating the MAC does not have the ability (based solely on this MAC construction) to ensure that the MAC it is producing is for the plaintext+metadata that the verifier will be asked to verify.  If the party generating the MAC had an authenticated assurance that this CID (including its length) is the one that it is supposed to be using, this would be okay (because that is the extra information needed to know that the MAC being produced is over the message that the verifier will be asked to verify), but I have demonstrated a situation where the client does not have such an authenticated assurance.

In short, you asked me to show how having a cid-length (in a different position than currently) will prevent an attack: the attack in question occurs when the client is generating a (its first) MAC, not at the time when the MAC is validated.

**Comment by boaks** [comment]
> the attack in question occurs when the client is generating a (its first) MAC, not at the time when the MAC is validated.

I still fail to follow your description, I can't see the attack.

**Comment by boaks** [comment]
The original topic to clarify the considerations for "variable cid length" is specified in the links below  

[variable length cid - responsible for constructing](https://github.com/tlswg/dtls-conn-id/blob/master/draft-ietf-tls-dtls-connection-id.md#L137)

[variable length cid - self-delineating](https://github.com/tlswg/dtls-conn-id/blob/master/draft-ietf-tls-dtls-connection-id.md#L264)

(Sorry, the `#Lnnn` is not wokring for .md)

If attacks on the MAC validation with CIDs using ambiguous length ares still possible, seem to be left to everyone's own consideration. Without examples it's hardly to follow nor to estimate the effect of such an attack. I personally consider AEAD, AES-CBC MtE not to be effected by such a "modified SERVER_HELLO" attack. AES-CBC EtM the validation of the clients  FINISHs MAC would fail to detect the modification (as some other modifications of dtls 1.2 handshake messages would also not be detected). But it's hard to see how that should break more, because the next steps will hardly be passed (decrypt => result != FINISH, verify_data would hardly match).
 
Maybe recommending to use [RFC7627 -  Extended Master Secret Extension](https://tools.ietf.org/html/rfc7627) would also mitigate that without adapting the MAC.

**Comment by jsalowey** [comment]
I don't think we have closed on the issue yet.   The fact that there is ambiguity in the MAC calculation is enough to cause concern.    We don't want to wait for a full break to be found, we should fix this now before publication.

**Comment by kaduk** [comment]
I agree that we have not closed on the MAC/AAD construction yet.
That said, I think we have used this github issue for discussion of multiple topics over its lifetime (probably my fault as I tried to keep up with the move from email to github -- sorry about that!).  The stated topic of the issue as to variable-length CID encoding being deterministic is resolved, in that the party that will receive packets using that CID has to use a deterministic/self-delineating encoding.  The party that will send packets does not need to know what that encoding is, though, and that is where the other issues we're talking about come into play.

**Comment by boaks** [comment]
Sorry, if I closed this to early.

For me:

> If, however, an implementation chooses to receive different lengths of CID, the assigned CID values must be self-delineating since there is no other mechanism available to determine what connection (and thus, what CID length) is in use. 

seems to be clear enough.

Though @kaduk denied, that this issue would also mitigate the MAC change, I assumed the MAC is decoupled for this and so I closed it. As I read @kaduk comment, we agree on closing this issue.

@jsalowey 

Do you think, it's worth to keep it open?

Or should we create a new topic and add all the MAC discussion e-mail to that?

**Comment by boaks** [comment]
The discussion about a new MAC is in PR #77 .

Therefore I close this again.
